static/js/mouse_util.js

```diff
@@ -0,0 +1,64 @@
+// function to calculate the drag start and end positions and time and
+// determine if the motion was a drag or a normal click.
+function drag() {
+    let start;
+    let time;
+
+    return {
+        start: function (e) {
+            start = { x: e.offsetX, y: e.offsetY };
+            time = new Date().getTime();
+        },
+
+        end: function (e) {
+            const end = { x: e.offsetX, y: e.offsetY };
+
+            let dist;
+            if (start) {
+                // get the linear difference between two coordinates on the screen.
+                dist = Math.sqrt(Math.pow(end.x - start.x, 2) + Math.pow(end.y - start.y, 2));
+            } else {
+                // this usually happens if someone started dragging from outside of
+                // a message and finishes their drag inside the message. The intent
+                // in that case is clearly to select an area, not click a message;
+                // setting dist to Infinity here will ensure that.
+                dist = Infinity;
+            }
+
+            this.val = dist;
+            this.time = new Date().getTime() - time;
+
+            start = undefined;
+
+            return dist;
+        },
+        is_short_and_quick: function () {
+            return !!(this.val < 5 && this.time < 150 || this.val < 2);
+        },
+        val: null,
+    };
+}
+
+const msgbox_drag_tracker = drag();
```

Comment:
Shouldn't this go inside the `on_non_dragged_click` function?  That way, every time we set up a new element, each element will have its own tracker.

Reply:
Also, the name `msgbox_drag_tracker` is too specific a name.  It should be more like `elem_mouse_tracker` or something.  But definitely don't say `msgbox`--that defeats the whole idea of generalizing this.